pkg/comp-functions/functions/vshnnextcloud/collabora.go

```diff
@@ -0,0 +1,732 @@
+		}
+
+		if val, ok := annotations["cert-manager.io/cluster-issuer"]; ok {
+			ingress.Annotations["cert-manager.io/cluster-issuer"] = val
+		}
+	}
+
+	return svc.SetDesiredKubeObject(ingress, comp.GetName()+"-collabora-code-ingress", runtime.KubeOptionAddLabels(labelMap))
+}
+
+func createIssuer(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
+	issuer := &cmv1.Issuer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      comp.GetName() + "-collabora-code-issuer",
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Spec: cmv1.IssuerSpec{
+			IssuerConfig: cmv1.IssuerConfig{
+				SelfSigned: &cmv1.SelfSignedIssuer{},
+			},
+		},
+	}
+
+	return svc.SetDesiredKubeObject(issuer, comp.GetName()+"-collabora-code-issuer", runtime.KubeOptionAddLabels(labelMap))
+}
+
+func createCertificate(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
+	certificate := &cmv1.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      comp.GetName() + "-collabora-code-certificate",
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Spec: cmv1.CertificateSpec{
+			SecretName: "collabora-code-tls",
+			DNSNames:   []string{comp.Spec.Parameters.Service.Collabora.FQDN},
+			IssuerRef: certmgrv1.ObjectReference{
+				Name: comp.GetName() + "-collabora-code-issuer",
+			},
+		},
+	}
+
+	return svc.SetDesiredKubeObject(certificate, comp.GetName()+"-collabora-code-certificate", runtime.KubeOptionAddLabels(labelMap))
+}
+
+func createCoolWSDConfigMap(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "collabora-code-coolwsd-config",
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Data: map[string]string{
+			"coolwsd.xml": coolwsd,
+			"coolkit.xml": coolkit,
+		},
+	}
+
+	return svc.SetDesiredKubeObject(cm, comp.GetName()+"-collabora-code-coolwsd-config", runtime.KubeOptionAddLabels(labelMap))
+}
+
+func createSecretWithRSAKeys(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
+	objName := comp.GetName() + "-collabora-code-ssh-key"
+
+	currentSecret := &corev1.Secret{}
+
+	err := svc.GetObservedKubeObject(currentSecret, objName)
+	if err == nil {
+		/*
+		 this part might look weird, but it's necessary because:
+		 - we don't want to overwrite the secret if it already exists
+		 - RSA generation is really slow
+		*/
+		return svc.SetDesiredKubeObject(currentSecret, objName, runtime.KubeOptionAddLabels(labelMap))
+	}
+
+	privkey, pubkey, err := generateSSHRSAKey()
+	if err != nil {
+		return err
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      objName,
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Data: map[string][]byte{
+			"id_rsa":     []byte(privkey),
+			"id_rsa.pub": []byte(pubkey),
+		},
+	}
+
+	return svc.SetDesiredKubeObject(secret, objName, runtime.KubeOptionAddLabels(labelMap))
+}
+
+func createServiceAccount(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
```

Comment:
There's `common.AddSaWithRole()` that can create an SA, role and binding for you, neatly packed into one handy function.